app/src/main/java/com/techcourse/DispatcherServlet.java

```diff
@@ -43,20 +45,24 @@ protected void service(final HttpServletRequest request, final HttpServletRespon
             final ModelAndView modelAndView = handlerAdapter.handle(request, response, mappedHandler);
             final View view = modelAndView.getView();
             move(view.getName(), request, response);
-
-        } catch (Exception e) { // TODO: 세분화
-            log.error("Exception : {}", e.getMessage(), e);
-            throw new ServletException(e.getMessage());
+        } catch (Exception e) {
+            log.error("exception occurred : {}", e.getMessage(), e);
+        } catch (Error e) {
+            log.error("error occurred : {}", e.getMessage(), e);
```

Comment:
이건 진짜 궁금해서 여쭤봐요! Exception 과 Error가 어떻게 다른가요?
두 예외를 세분화하신 이유가 궁금하네요

Reply:
Throwable의 서브타입은 크게 Error, Exception으로 나뉩니다!

Error의 경우에는 SOF와 같은 복구 불가능한 심각한 상황을 나타냅니다.
대부분 시스템 자원 부족 등 저수준에서 발생하는 예외 상황을 Error로 나타내는 것 같습니다.

기존 코드에서는 Throwable로만 Catch하고 Exception으로 전환하더라고요.

굳이 에러 상황까지 같이 Catch를 해서 예외로 전환할 필요는 없다고 생각해 분리했습니다.

그런데 답글달면서 찾아보다가 Error의 JavaDocs를 보니까 다음과 같은 문구가 있네요 ㅋㅋ

> An Error is a subclass of Throwable that indicates serious problems that a reasonable application should not try to catch. 

Error를 catch하는 구문은 없어도 되지 않을까.. 싶네요! 그런데 이런 경우 로깅이 불가능할 것 같은데 후추는 어떻게 생각하세요? 둘 중 어떤 방식이 나을까요?

1. 표준에서 권장하는 방식은 아니지만 로깅을 위해 Error의 서브타입도 catch한다.
2. 표준을 따르고 Error에 대해서는 로깅을 하지 않는다.

Reply:
어려운 문제네요.

조심스럽게 생각해보기로는, 저는 `로깅을 위해 error의 서브타입도 catch 할 수 있다` 쪽에 가깝습니다.

javadoc 의 내용을 더 읽어봤습니다.

인용해주신 문장 바로 다음에 다음과 같은 문장이 있네요. 문맥상 이 문장이 `A reasonable application should not try to catch( an error.)`의 근거겠네요.

> Most such errors are abnormal conditions.

결국 비정상적인 조건에 의해 error 가 발생하기 때문에, 이를 catch 할 생각을 하지 말라는 것으로 추측됩니다.

그러나 Error의 하위타입 중 하나인 ThreadDeath 의 javadoc에는 다음과 같은 문장이 있습니다.

> An application should catch instances of this class only if it must clean up after being terminated asynchronously. If ThreadDeath is caught by a method, it is important that it be rethrown so that the thread actually dies.

저는 이러한 뉘앙스를 봤을 때, `절대 아무것도! catch 조차! 하지 말아라` 라기 보다는, `Exception을 다룰 때처럼 catch 해서 별도의 로직을 수행하는 등의 자유로운 예외 처리를 하지 말아라` 정도의 의미라고 생각합니다.

만약 정말 catch 를 하면 안 됐다면, javadoc의 표현도 더 강했을 것 같아요. should가 아닌 must, have to 정도로요.

결론은 Error 를 catch 해서 log를 남기는 정도는 괜찮을 것 같습니다. 다만, 주석으로 왜 이 Error를 잡았는지 정도를 적어두면 좋을 것 같습니다!

제 생각을 여쭤봐주셔서 감사합니다 👍  좋은 공부가 됐어요 😄

Reply:
직접 ThreadDeathError까지 확인해보셨군요..!
말씀주신 대로 '무조건 잡지 말아야 한다' 가 아니라 특정한 이유가 있다면 잡아도 되는 것 같습니다.

저는 무조건 표준 문서만 보고 '아 잡지 말아야 되는거구나' 생각했었는데 역시 생각이 유연하시군요 👍👍

저도 말씀해주신 부분에 동감해서 기존 코드를 유지시킬 생각입니다! 감사합니다!